### ISSUE TITLE
[0.4.5] restrict-plus-operands misses any returned by calls, indexed array elements, and compound assignments

### ISSUE THREAD

**Issue by fireairforce** [utooland/utoo-lint#1277]
## Summary

With explicit allowAny:false, the rule misses function calls returning any, any[] element accesses and += assignments. An explicit skipCompoundAssignments:false still misses the compound case. Direct any operands are correctly reported; corresponding number controls stay clean. These examples do not use object destructuring and are separate from #1273.

Verified with @utoo/lint 0.4.5, ESLint 8.57.1, @typescript-eslint/parser and eslint-plugin 6.21.0, eslint-plugin-react 7.37.5, Node.js 22.23.1 on macOS arm64. Each example enables only the indicated rule, with no parsing failures or suppression comments.

## Setup

```sh
npm init -y
npm install --save-dev @utoo/lint@0.4.5 eslint@8.57.1 @typescript-eslint/parser@6.21.0 @typescript-eslint/eslint-plugin@6.21.0 eslint-plugin-react@7.37.5 typescript@5.3.3
```

`tsconfig.json`:

```json
{"compilerOptions":{"strict":true,"target":"ES2022","jsx":"react"},"include":["example.tsx"]}
```

Run each case separately, replacing example.tsx and both config files with the corresponding contents below:

```sh
./node_modules/.bin/eslint --no-eslintrc --config .eslintrc.cjs --format=json example.tsx
./node_modules/.bin/utlint --config utlint.config.json --format=json example.tsx
```

## return-any

`example.tsx`:

```tsx
function read():any{return 1;}export const result=read()+1;
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number or string, allowing a string + any of: `boolean`, `null`, `RegExp`, `undefined`. Got `any`.
```

utlint messages:

```text
(none)
```

## plus-generic-any

`example.tsx`:

```tsx
function read<T>(x:T):T{return x;}export function example(x:any){return read(x)+1;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number or string, allowing a string + any of: `boolean`, `null`, `RegExp`, `undefined`. Got `any`.
```

utlint messages:

```text
(none)
```

## array-any

`example.tsx`:

```tsx
export function example(x:any[]){return x[0]+1;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number or string, allowing a string + any of: `boolean`, `null`, `RegExp`, `undefined`. Got `any`.
```

utlint messages:

```text
(none)
```

## compound-True

`example.tsx`:

```tsx
export function example(x:any){let result=1;result+=x;return result;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false,
        "skipCompoundAssignments": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false,
        "skipCompoundAssignments": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number or string, allowing a string + any of: `boolean`, `null`, `RegExp`, `undefined`. Got `any`.
```

utlint messages:

```text
(none)
```

## return-number

`example.tsx`:

```tsx
function read():number{return 1;}export const result=read()+1;
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```

## array-number

`example.tsx`:

```tsx
export function example(x:number[]){return x[0]+1;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```

## assignment-number

`example.tsx`:

```tsx
export function example(x:number){let result=1;result+=x;return result;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```

## direct-any

`example.tsx`:

```tsx
export function example(x:any){return x+1;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 1 diagnostic(s). utlint: 1 diagnostic(s).

eslint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number or string, allowing a string + any of: `boolean`, `null`, `RegExp`, `undefined`. Got `any`.
```

utlint messages:

```text
Invalid operand for a '+' operation. Operands must each be a number, string, or bigint. Got `any`.
```

## compound-False

`example.tsx`:

```tsx
export function example(x:any){let result=1;result+=x;return result;}
```

`.eslintrc.cjs`:

```js
module.exports = {
  "root": true,
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2022,
    "sourceType": "module",
    "ecmaFeatures": {
      "jsx": true
    },
    "project": "./tsconfig.json"
  },
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false,
        "skipCompoundAssignments": true
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
};
```

`utlint.config.json`:

```json
{
  "rules": {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        "allowAny": false,
        "skipCompoundAssignments": true
      }
    ]
  },
  "settings": {
    "react": {
      "version": "18.2"
    }
  }
}
```

ESLint: 0 diagnostic(s). utlint: 0 diagnostic(s).

eslint messages:

```text
(none)
```

utlint messages:

```text
(none)
```
